specs/LaunchingRegisteredProtocols.md

```diff
@@ -0,0 +1,264 @@
+                    // popup dialog.
+                }
+                else
+                {
+                    CHECK_FAILURE(args->put_Handled(FALSE));
+                    // Otherwise allow the popup dialog, and allow the user to decide
+                    // whether or not to allow the protocol to launch.
+                }
+            return S_OK;
+        })
+        .Get(),
+    &m_launchingRegisteredProtocolToken));
+
+CHECK_FAILURE(frame->add_LaunchingRegisteredProtocol(
+    Callback<ICoreWebView2FrameLaunchingRegisteredProtocolEventHandler>(
+        [this](
+            ICoreWebView2Frame* sender,
+            ICoreWebView2LaunchingRegisteredProtocolEventArgs* args) {
+                wil::unique_cotaskmem_string uri;
+                CHECK_FAILURE(args->get_Uri(&uri));
+                if (wcscmp(uri.get(), L"calculator://") == 0)
+                {
+                    // If this matches our desired protocol, then suppress the
+                    // popup dialog.
+                    CHECK_FAILURE(args->put_Handled(TRUE));
+                }
+                else
+                {
+                    // Otherwise revoke permissions previously granted to this protocol 
+                    // from this origin. This will trigger the dialog to appear to allow the user
+                    // to confirm or deny the protocol launch.
+                    CHECK_FAILURE(args->put_RevokeProtocolPermissionsPerOrigin(TRUE));
+                }
+                return S_OK;
+        })
+        .Get(),
+    &m_frameLaunchingRegisteredProtocolToken));
+    
+``` 
+
+## .NET and WinRT
+
+```c #
+
+void WebView_LaunchingRegisteredProtocol(object target, CoreWebView2LaunchingRegisteredProtocolEventArgs e)
+{
+    if (e.Uri == "calculator:///") 
+    {
+        // If this matches our desired protocol, then suppress the popup dialog.
+        e.Handled = true;
+    }
+    else 
+    {
+        // Otherwise allow the popup dialog, and allow the user to decide 
+        // whether or not to allow the protocol to launch.
+        e.Handled = false;
+    }
+    webView.CoreWebView2.FrameCreated += (sender, args) =>
+    {
+        // Apply the same logic as above to non-main frame raising the event.
+        args.Frame.LaunchingRegisteredProtocol += (frameSender, LaunchingRegisteredProtocolArgs) =>
+        {
+            if (LaunchingRegisteredProtocolArgs.Uri == "calculator:///") 
+            {
+                // If this matches our desired protocol, then suppress the popup dialog.
+                e.Handled = true;
+            }
+            else 
+            {
+                // Otherwise revoke permissions previously granted to this protocol 
+                // from this origin. This will trigger the dialog to appear to allow the user
+                // to confirm or deny the protocol launch.
+                e.RevokeProtocolPermissionsPerOrigin = true;
```

Comment:
This means we will always be ignoring the end user checking the box to not show the dialog in the future right? Is the purpose of this flag to treat the current event arg's origin as unstrusted and to force showing the dialog? If so, should we also not be showing the checkbox in the dialog? 

Would it be easier to have a bool setting somewhere to treat all origins as unstrusted? Or a list in settings of origins to treat as untrusted?

I don't think I understand what the scenario for this property is

Reply:
Office was asking for a way to previously revoke permissions given in the past, as once the checkbox is clicked there isn't a way to uncheck this option. If the end dev sets this option and does not change this boolean back, we are essentially ignoring the end user checking the box. Correct, this will force showing the dialog. 
I think changing this to a setting that gives the option to disable the checkbox makes the most sense to me.

Reply:
Spoke to Limin and decided to change this to a setting that disables the checkbox/revokes previous permissions.